Userland/Utilities/run-tests.cpp

```diff
@@ -0,0 +1,327 @@
+        }
+        print_modifiers({ CLEAR });
+    }
+
+    // Make sure our clear modifiers goes through before we dump file output via write(2)
+    fflush(stdout);
+
+    if (print_stdout_stderr && test_result.stdout_err_fd > 0) {
+        int ret = lseek(test_result.stdout_err_fd, 0, SEEK_SET);
+        VERIFY(ret == 0);
+        for (;;) {
+            char buf[32768];
+            ssize_t nread = read(test_result.stdout_err_fd, buf, sizeof(buf));
+            if (nread == 0)
+                break;
+            if (nread < 0) {
+                perror("read");
+                break;
+            }
+            size_t already_written = 0;
+            while (already_written < (size_t)nread) {
+                ssize_t nwritten = write(STDOUT_FILENO, buf + already_written, nread - already_written);
+                if (nwritten < 0) {
+                    perror("write");
+                    break;
+                }
+                already_written += nwritten;
+            }
+        }
+    }
+
+    close(test_result.stdout_err_fd);
+}
+
+FileResult TestRunner::run_test_file(const String& test_path)
+{
+    double start_time = get_time_in_ms();
+
+    auto path_for_test = LexicalPath(test_path);
+    if (should_skip_test(path_for_test)) {
+        return FileResult { move(path_for_test), 0.0, Test::Result::Skip, -1 };
+    }
+
+    // FIXME: actual error handling, mark test as :yaksplode: if any are bad instead of VERIFY
+    posix_spawn_file_actions_t file_actions;
+    posix_spawn_file_actions_init(&file_actions);
+    char child_out_err_path[] = "/tmp/run-tests.XXXXXX";
+    int child_out_err_file = mkstemp(child_out_err_path);
+    VERIFY(child_out_err_file >= 0);
+
+    int ret = unlink(child_out_err_path);
+    VERIFY(ret == 0);
+
+    (void)posix_spawn_file_actions_adddup2(&file_actions, child_out_err_file, STDOUT_FILENO);
+    (void)posix_spawn_file_actions_adddup2(&file_actions, child_out_err_file, STDERR_FILENO);
+    (void)posix_spawn_file_actions_addchdir(&file_actions, path_for_test.dirname().characters());
+
+    Vector<const char*, 4> argv;
+    argv.append(path_for_test.basename().characters());
+    auto extra_args = m_config->read_entry(path_for_test.basename(), "Arguments", "").split(' ');
+    for (auto& arg : extra_args)
+        argv.append(arg.characters());
+    argv.append(nullptr);
+
+    pid_t child_pid = -1;
+    // FIXME: Do we really want to copy test runner's entire env?
+    ret = posix_spawn(&child_pid, test_path.characters(), &file_actions, nullptr, const_cast<char* const*>(argv.data()), environ);
+    VERIFY(ret == 0);
+    VERIFY(child_pid > 0);
+
+    int wstatus;
+    ret = waitpid(child_pid, &wstatus, 0);
+    VERIFY(ret == child_pid);
+
+    Test::Result test_result = Test::Result::Pass;
+    if (WIFEXITED(wstatus)) {
+        if (wstatus != 0) {
+            test_result = Test::Result::Fail;
+        }
+    } else if (WIFSTOPPED(wstatus)) {
+        VERIFY_NOT_REACHED(); // what? why?
```

Comment:
Program sent itself a `SIGSTOP` or something happened with `SIGTTOU`?
Should probably just send a `SIGCONT` and call the test bad if it fails again

Reply:
Sure, I added a simple for loop that tries to waitpid a few times and if it keeps coming back WIFSTOPPED it bails with Test::Result::Fail